File: p5/c/p5.c
/* IMPORT LIBRARIES */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/* GLOBAL VARIABLES */
#define limit 10
int dividers[limit];

/* FUNCTION DECLARATIONS */
int is_prime(int);
void factorize(int, int*);

/* MAIN PROGRAM */
int main(void)
{
	int i, j;
	for(i = 2; i <= limit; i++)
	{
		int* factors = (int*) malloc(sizeof(int));
		factorize(i, factors);

		int count;
		int elem;
		j = 1;
		while(j <= factors[0])
		{
			count = 1;			
			elem = factors[j];
			while((j < factors[0]))
			{
				if(factors[j + 1] == elem)
				{
					count++;
					j++;
				} else
				{
					j++;
				}
			}
			if(count > dividers[elem - 1])
			{
				dividers[elem - 1] = count;
			}
			if(j == factors[0]) {j++;}
		}
		free(factors);
	}

	int number = 1;
	int exponent;
	for(i = 0; i < limit; i++)
	{
		exponent = dividers[i];
		if ( exponent != 0)
			number *= pow(i + 1, exponent);
	}

	printf("Smallest number that can be divided for all [1, %d] is %d\n", limit, number);
	return 0;
}

int is_prime(int x)
{
	if(x == 1)
		return 0;
	if ((x == 2) || (x == 3))
		return 1;
	if (x % 2 == 0)
		return 0;

	int check = 1;
	int i;
	int sqrt_x = sqrt(x);
	for(i = 3; i <= sqrt_x; i += 2)
	{
		if (x % i == 0)
			check = 0;
	}

	return check;
}

void factorize(int x, int* arr)
{
	if ((x == 2) || (x == 3))
	{
		arr[0]++;
		arr = (int*) realloc(arr, (arr[0] + 1) * sizeof(int));
		arr[arr[0]] = x;
	} else
	{
		int i;
		int sqrt_x = sqrt(x);

		for (i = 2; i <= sqrt_x; i++)
		{
			while(x % i == 0)
			{
				arr[0]++;
				arr = (int*) realloc(arr, (arr[0] + 1) * sizeof(int));
				arr[arr[0]] = i;
				x /= i;
			}
		}

		if (is_prime(x))
		{
			arr[0]++;
			arr = (int*) realloc(arr, (arr[0] + 1) * sizeof(int));
			arr[arr[0]] = x;
		}
	}
}

File: p5/python/p5.py
import numpy as np

limit = 10
dividers = np.zeros(limit)

def is_prime(x):
    if (x == 1):                            return False
    if (x == 2) or (x == 3) or (x == 5):    return True
    if (x % 2 == 0):                        return False

    check = True
    for i in range(3, int(np.sqrt(x)) + 1, 2):
        if (x % i == 0):
            check = False
    
    return check

def factorize(x):
    if (x == 1): return []
    if (x == 2) or (x == 3): return [x]

    factors = np.array([])
    i = 2
    while i < (int(np.sqrt(x)) + 1):
        if (x % i == 0):
            if is_prime(i):
                factors = np.append(factors, i)
                factors = np.append(factors, factorize(x / i))
                break
            else:
                factorize(i)
        else:
            i += 1
    
    if is_prime(x): factors = np.append(factors, x)
    return factors.astype(int)


if __name__ == "__main__":
    for i in range(2, limit + 1):
        factors = factorize(i)
        elems, counts = np.unique(factors, return_counts = True)
        for j in range(len(elems)):
            if (counts[j] > dividers[elems[j] - 1]):
                dividers[elems[j]  - 1] = counts[j]

    number = 1
    for i in range(limit):
        exponent = dividers[i]
        if exponent != 0: number *= (i+1)**exponent
    print("Smallest number that can be divided for all [1," + str(limit) +"] is:", int(number))

